src/NLog/Internal/MruCache.cs

```diff
@@ -0,0 +1,185 @@
+        public MruCache(int maxCapacity)
+        {
+            _maxCapacity = maxCapacity;
+            _dictionary = new Dictionary<TKey, MruItem>(_maxCapacity);
+            _currentVersion = 1;
+        }
+
+        /// <summary>
+        /// Insert item into caceh
+        /// </summary>
+        /// <param name="key">Dictionary Key</param>
+        /// <param name="value">Dictionary Value</param>
+        /// <returns>Key did not exist already</returns>
+        public bool TryAddValue(TKey key, TValue value)
+        {
+            lock (_dictionary)
+            {
+                MruItem item;
+                if (_dictionary.TryGetValue(key, out item))
+                {
+                    var version = _currentVersion;
+                    if (item.Version != version || !EqualityComparer<TValue>.Default.Equals(value, item.Value))
+                    {
+                        _dictionary[key] = new MruItem(value, version);
+                    }
+                    return false;   // Already exists
+                }
+                else
+                {
+                    var version = ++_currentVersion;
+                    if (_dictionary.Count >= _maxCapacity)
+                    {
+                        // Make 3 sweeps:
+                        //  1) Versions below _currentVerison - _maxCapacity / 2
+                        //  2) Versions below _currentVerison - _maxCapacity / 10
+                        //  3) Versions below _currentVerison - 2
+                        var pruneKeys = new List<TKey>(_dictionary.Count);
+                        for (int i = 1; i <= 3; ++i)
+                        {
+                            long oldVersion = version - 2;
+                            switch (i)
+                            {
+                                case 1: oldVersion = version - (int)(_maxCapacity / 1.5); break;
+                                case 2: oldVersion = version - _maxCapacity / 10; break;
+                            }
+                            foreach (var element in _dictionary)
+                            {
+                                if (element.Value.Version <= oldVersion)
+                                {
+                                    pruneKeys.Add(element.Key);
+                                    if (_dictionary.Count - pruneKeys.Count < _maxCapacity / 1.5)
+                                    {
+                                        i = 3;
+                                        break;  // Do not clear the entire cache
+                                    }
+                                }
+                            }
+                        }
+
+                        foreach (var pruneKey in pruneKeys)
+                        {
+                            _dictionary.Remove(pruneKey);
+                        }
+
+                        if (_dictionary.Count >= _maxCapacity)
+                        {
+                            _dictionary.Clear();
```

Comment:
The `_dictionary.Clear();` is a fail safe, in case something has gone terrible wrong, and all items has the same version.